src/node/txdownload_impl.cpp

```diff
@@ -0,0 +1,415 @@
+            // from any of our non-wtxidrelay peers.
+            m_recent_rejects.insert(tx->GetHash());
+            m_recent_rejects.insert(tx->GetWitnessHash());
+            m_txrequest.ForgetTxHash(tx->GetHash());
+            m_txrequest.ForgetTxHash(tx->GetWitnessHash());
+            return false;
+        }
+        return true;
+    }
+    case TxValidationResult::TX_INPUTS_NOT_STANDARD:
+    {
+        // If the transaction failed for TX_INPUTS_NOT_STANDARD,
+        // then we know that the witness was irrelevant to the policy
+        // failure, since this check depends only on the txid
+        // (the scriptPubKey being spent is covered by the txid).
+        // Add the txid to the reject filter to prevent repeated
+        // processing of this transaction in the event that child
+        // transactions are later received (resulting in
+        // parent-fetching by txid via the orphan-handling logic).
+        if (tx->GetWitnessHash() != tx->GetHash()) {
+            m_recent_rejects.insert(tx->GetHash());
+            m_txrequest.ForgetTxHash(tx->GetHash());
+        }
+        break;
+    }
+    case TxValidationResult::TX_CONSENSUS:
+    case TxValidationResult::TX_RECENT_CONSENSUS_CHANGE:
+    case TxValidationResult::TX_NOT_STANDARD:
+    case TxValidationResult::TX_PREMATURE_SPEND:
+    case TxValidationResult::TX_WITNESS_MUTATED:
+    case TxValidationResult::TX_CONFLICT:
+    case TxValidationResult::TX_MEMPOOL_POLICY:
+        break;
+    }
+    // We can add the wtxid of this transaction to our reject filter.
+    m_recent_rejects.insert(tx->GetWitnessHash());
+    // Forget requests for this wtxid, but not for the txid, as another version of
+    // transaction may be valid. No-op if the tx is not in txrequest.
+    m_txrequest.ForgetTxHash(tx->GetWitnessHash());
+    // If it came from the orphanage, remove it (this doesn't happen if the transaction was missing
+    // inputs). No-op if the tx is not in the orphanage.
+    m_orphanage.EraseTx(tx->GetWitnessHash());
+    m_orphan_resolution_tracker.ForgetTxHash(tx->GetWitnessHash());
+    return false;
+}
+
+bool TxDownloadImpl::AlreadyHaveTxLocked(const GenTxid& gtxid) const
+    EXCLUSIVE_LOCKS_REQUIRED(m_tx_download_mutex)
+{
+    const uint256& hash = gtxid.GetHash();
+
+    if (m_orphanage.HaveTx(gtxid)) return true;
+
+    if (m_recent_confirmed_transactions.contains(hash)) return true;
+
+    return m_recent_rejects.contains(hash) || m_opts.m_mempool_ref.exists(gtxid);
+}
+bool TxDownloadImpl::AlreadyHaveTx(const GenTxid& gtxid) const
+    EXCLUSIVE_LOCKS_REQUIRED(!m_tx_download_mutex)
+{
+    LOCK(m_tx_download_mutex);
+    return AlreadyHaveTxLocked(gtxid);
+}
+
+void TxDownloadImpl::AddTxAnnouncement(NodeId peer, const GenTxid& gtxid, std::chrono::microseconds now)
+    EXCLUSIVE_LOCKS_REQUIRED(m_tx_download_mutex)
+{
+    if (!Assume(m_peer_info.count(peer) > 0)) return;
+    if (m_orphanage.HaveTx(gtxid)) AddOrphanAnnouncer(peer, gtxid.GetHash(), now);
```

Comment:
this adds an announcer even if the INV is txid type, even though the assumption is wtxid arg

Reply:
I think the simplest solution is just filter this addition based on wtxid-ness, since this only helps in the multi-announcer case, which wouldn't be a regression, iiuc

Reply:
ooh thanks :+1:

Reply:
Fixed, just did a gate on IsWtxid. Don't think that many non-wtxidrelay nodes are still out there anyway.